### PR TITLE
osu! speedruns concept poc

### PR DESCRIPTION
**Pull request by cmyui**
just meme w/ justice... for now

```sql
create table user_speedruns (
    id int not null primary key,
    user_id int not null,
    game_mode int not null,
    timeframe varchar(32) not null,
    score_type varchar(16) not null,
    score_value int not null,
    started_at datetime not null,
    ended_at datetime null,
    cancelled_at datetime null
);

alter table user_speedruns add index (user_id);
alter table user_speedruns add index (game_mode);
alter table user_speedruns add index (timeframe);
alter table user_speedruns add index (score_type);
```

### CODE REVIEW

**Comment by cmyui** [comment]
meme for now but maybe real one day

File: constants/chatbotCommands.py [hunk 1]
```diff
@@ -98,6 +100,99 @@ def wrapper(f: CommandCallable) -> CommandCallable:
...
+        user_id=user_id,
+        game_mode=game_mode,
+        score_type=score_type,
+        timeframe=timeframe,
+    )
+    if not user_speedruns:
+        return "You must first run"
+
+    best_speedrun = max(user_speedruns, key=lambda s: s.score_value)
+    return f"Your best speedrun: {best_speedrun.score_value:,.2f} in {best_speedrun.timeframe}"
```
Comment (cmyui) [inline constants/chatbotCommands.py:201]:
```suggestion
    # TODO: show scores under this like !speedrun end
    return f"Your best speedrun: {best_speedrun.score_value:,.2f} in {best_speedrun.timeframe}"
```

File: common/speedrunning.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+class UserSpeedrun:
+    id: UUID
+    user_id: int
+    game_mode: int
+    timeframe: SpeedrunTimeframe
+    score_type: ScoreType
+    score_value: int
+    started_at: datetime
+    ended_at: datetime | None
+    cancelled_at: datetime | None
```
Comment (cmyui) [inline common/speedrunning.py:39]: todo rm `cancelled_at`, can virtually figure this out `if ended_at and (ended_at - started_at) < timeframe`

**Comment by cmyui** [comment]
@tsunyoku thoughts on merging this/considering building it out? Might be a fun thing to build out more proper support for (e.g. leaderboards on new website, etc.)

**Comment by cmyui** [comment]
Nw if not

**Comment by tsunyoku** [comment]
yeah happy with the concept existing, would be better to understand how you envision it working though

**Comment by cmyui** [comment]
> yeah happy with the concept existing, would be better to understand how you envision it working though

The original idea was from Justice but I had a couple of ideas -- we can start a draft

File: .github/workflows/production-deploy.yml [hunk 3]
```diff
@@ -4,6 +4,7 @@ on:
   push:
     branches:
       - main
+      - speedruns
```
Comment (cmyui) [inline .github/workflows/production-deploy.yml]:
```suggestion
```

File: constants/chatbotCommands.py [hunk 4]
```diff
@@ -110,6 +111,100 @@ def wrapper(f: CommandCallable) -> CommandCallable:
...
+    )
+
+    return f"Speedrun started! ID: {user_speedrun.id}"
+
+
+@command(trigger="!speedrun end", hidden=False)
+async def end_speedrun(fro: str, chan: str, message: list[str]) -> str:
+    user_id = await user_utils.get_id_from_username(fro)
+    # if user_id not in {1001, 4528}:
+    #     return "u may not taste my delicious lettuce"
```
Comment (cmyui) [inline constants/chatbotCommands.py]:
```suggestion
```

File: constants/chatbotCommands.py [hunk 5]
```diff
@@ -110,6 +111,100 @@ def wrapper(f: CommandCallable) -> CommandCallable:
...
+        ret += f"{get_url_embed(beatmap_url, score.song_name)}: {score.value:,.2f}{mods_str}\n"
+
+    return ret[:-1]
+
+
+@command("!speedrun best", hidden=False)
+async def best_speedrun(fro: str, chan: str, message: list[str]) -> str:
+    user_id = await user_utils.get_id_from_username(fro)
+    # if user_id not in {1001, 4528}:
+    #     return "u may not taste my delicious lettuce"
```
Comment (cmyui) [inline constants/chatbotCommands.py]:
```suggestion
```

**Pull request merged**